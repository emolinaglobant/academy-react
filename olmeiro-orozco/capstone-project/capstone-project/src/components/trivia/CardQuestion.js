import { React, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import Swal from "sweetalert2";

import { increasePrize, addRound, resetGame, roundChange } from "../../actions/game";
import { getRandom } from "../../helpers/getRandom";

export const CardQuestion = () => {
  const allCategories = document.querySelectorAll(".category");

  const [numberQuestionAsked, setNumberQuestionAsked] = useState(0);
  const [questionIndex, setQuestionIndex] = useState(0);
  const [prize, setPrize] = useState(1);
  const [round, setRound] = useState(1);
  const [startGame, setStart] = useState(false);
  const [disabled, setDisabled] = useState(false);

  const dispatch = useDispatch();
  const { questions, start, changeRound } = useSelector((state) => state.game);
  
  useEffect(() => {
    setStart(start);
    setDisabled(changeRound);
    dispatch(addRound(round));
  }, [questions, round, prize, dispatch, changeRound]);

  const checkAnswer = (e) => {
    let answer = e.target.dataset.answer;
    if (answer === "correct_answer") {
      if (prize % 5 === 0 && numberQuestionAsked !== 1 && prize !== 25) {
        setNumberQuestionAsked(1);
        setQuestionIndex(0);
        setPrize(prize + 1);
        setRound(round + 1);

        setDisabled(true);
        dispatch(roundChange(true));

        dispatch(increasePrize(prize));
        dispatch(addRound(round));

        Swal.fire({
          icon: "success",
          title: "Excelent...",
          text: "Please choose another category",
        });
      } else if (prize % 5 === 0 && prize === 25) {
        const imagePath = "/assets/trivia2.jpg";
        Swal.fire({
          title: 'Triviaaa!',
          text: 'You are amazing.',
          imageUrl: `${imagePath}`,
          imageWidth: 400,
          imageHeight: 200,
          imageAlt: 'Custom image',
        })
        setStart(false);
        setQuestionIndex(0);
        setNumberQuestionAsked(1);
        setPrize(1);

        allCategories.forEach((btn) => {
          btn.removeAttribute("disabled");
        });
        dispatch(resetGame());
      } else {
        
        setNumberQuestionAsked(numberQuestionAsked + 1);
        setQuestionIndex(questionIndex + 1);
        setPrize(prize + 1);
        dispatch(increasePrize(prize));

        Swal.fire({
          icon: "success",
          title: "Excellent...",
          text: "Keep playing, click on the next question.",
        });
      }
    } else {
      handleErrorQuestion();
    }
  };

  const handleErrorQuestion = () => {
    Swal.fire({
      icon: "error",
      title: "Oops...",
      text: "Nice try!",
      footer: 'Try it again, you can do it!',
    });
    setStart(false);
    setQuestionIndex(0);
    setNumberQuestionAsked(1);
    setPrize(1);

    allCategories.forEach((btn) => {
      btn.removeAttribute("disabled");
    });
    dispatch(resetGame());
  };

  const handleRetire = () => {
    Swal.fire({
      icon: "error",
      title: "You won accumulated!",
      text: "Try it again!",
      footer: "Choose another category",
    });
    setStart(false);
    setQuestionIndex(0);
    setNumberQuestionAsked(1);
    setPrize(1);

    dispatch(resetGame());

    allCategories.forEach((btn) => {
      btn.removeAttribute("disabled");
    });
  };

  const order = getRandom(0, 5);
  return (
    <>
      {startGame && (
        <article className="card__question">
          <header className="card__question-title">
            <h3>
              {questions === undefined
                ? "loading"
                : questions[questionIndex].question}
            </h3>
            <h5>choose one option, only one is true.</h5>
          </header>

          <section className="card__question-options">
            <section className="card__options">
              <section className={`card__option order-${order}`}>
                <button
                  type="button"
                  data-answer="correct_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].correct_answer}
                </button>
              </section>

              <section className="card__option order-1">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[0]}
                </button>
              </section>

              <section className="card__option order-2">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[1]}
                </button>
              </section>

              <section className="card__option order-3">
                <button
                  type="button"
                  data-answer="incorrect_answer"
                  className="btn btn__option btn-answer btn-option"
                  disabled={disabled}
                  onClick={checkAnswer}
                >
                  {questions === undefined
                    ? "Loading"
                    : questions[questionIndex].incorrect_answers[2]}
                </button>
              </section>
            </section>
            <footer className="card__footer">
              <button
                type="button"
                className="btn btn__retirement"
                id="retire"
                onClick={handleRetire}
              >
                I want to retire
              </button>
            </footer>
          </section>
        </article>
      )}

      <section
        className="card__start"
        style={{ display: start ? "none" : "flex" }}
      >
        <h1>Please choose a category to start.</h1>
      </section>
    </>
  );
};