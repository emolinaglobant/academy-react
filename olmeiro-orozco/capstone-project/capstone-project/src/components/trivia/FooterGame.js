import React from "react";

export const FooterGame = () => {
  return (
    <footer className="trivia__footer">
      <p>all rights reserved - by @olmeiro_orozco 2022</p>
    </footer>
  );
};