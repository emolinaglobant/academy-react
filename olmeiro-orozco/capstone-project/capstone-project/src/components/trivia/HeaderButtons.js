import React from 'react'

export const HeaderButtons = () => {
  return (
     <>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__warning btn__header-question">1</button>
          <button type="button" className="btn btn__warning btn__header-question">2</button>
          <button type="button" className="btn btn__warning btn__header-question">3</button>
          <button type="button" className="btn btn__warning btn__header-question">4</button>
          <button type="button" className="btn btn__warning btn__header-question">5</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__warning btn__header-question">6</button>
          <button type="button" className="btn btn__warning btn__header-question">7</button>
          <button type="button" className="btn btn__warning btn__header-question">8</button>
          <button type="button" className="btn btn__warning btn__header-question">9</button>
          <button type="button" className="btn btn__warning btn__header-question">10</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__success btn__header-question">11</button>
          <button type="button" className="btn btn__success btn__header-question">12</button>
          <button type="button" className="btn btn__success btn__header-question">13</button>
          <button type="button" className="btn btn__success btn__header-question">14</button>
          <button type="button" className="btn btn__success btn__header-question">15</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__success btn__header-question">16</button>
          <button type="button" className="btn btn__success btn__header-question">17</button>
          <button type="button" className="btn btn__success btn__header-question">18</button>
          <button type="button" className="btn btn__success btn__header-question">19</button>
          <button type="button" className="btn btn__success btn__header-question">20</button>
        </div>
        <div className='btn__header-warginbox'>
          <button type="button" className="btn btn__danger btn__header-question">21</button>
          <button type="button" className="btn btn__danger btn__header-question">22</button>
          <button type="button" className="btn btn__danger btn__header-question">23</button>
          <button type="button" className="btn btn__danger btn__header-question">24</button>
          <button type="button" className="btn btn__danger btn__header-question">25</button>
        </div>
     </>
  )
}