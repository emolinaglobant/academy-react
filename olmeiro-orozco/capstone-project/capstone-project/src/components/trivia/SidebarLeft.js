import React from 'react'
import { useDispatch, useSelector } from 'react-redux';

import { startRound, loadQuestions, roundChange } from "../../actions/game";
import { getDifficulty } from '../../helpers/getDifficulty';

export const SidebarLeft = () => {

  const dispatch = useDispatch();
  const {round, changeRound } = useSelector( state => state.game );

  const clickCategory = (e) => {
    if(changeRound === true){
      dispatch(roundChange(false));
    }
    e.target.disabled = true;
    let numCategory = e.target.dataset.numbercategory;
    const newDifficulty = getDifficulty(round);
    dispatch(startRound(true, numCategory, newDifficulty));
    dispatch(loadQuestions());
  }
 
  return (
    <aside className='trivia__sidebar-left'>
        <h2>Difficulty level</h2>
        <section className="trivia__difficulty">
          <button type="button" data-level="easy" className="btn btn__warning mr-3 ml-3">Low</button>
          <button type="button" data-level="medium" className="btn btn__success mr-3">Medium</button>
          <button type="button" data-level="hard" className="btn btn__danger mr-3">High</button>
        </section>

        <section className="trivia__categories">
          <h3>Categories</h3>
          <button onClick={ clickCategory }  type="button" className="btn btn__outline-dark mt-3 mr-3 category" data-category="Celebrities" data-numbercategory="26">Celebrities</button>
          <button onClick={ clickCategory }  type="button" className="btn btn__outline-dark mt-3 mr-3 category" data-category="History" data-numbercategory="23">History</button>
          <button onClick={ clickCategory }  type="button" className="btn btn__outline-dark mt-3 mr-3 category" data-category="Sports" data-numbercategory="21">Sports</button>
          <button onClick={ clickCategory }  type="button" className="btn btn__outline-dark mt-3 mr-3 category" data-category="Animals" data-numbercategory="27">Animals</button>
          <button onClick={ clickCategory }  type="button" className="btn btn__outline-dark mt-3 mr-3 category" data-category="Geography" data-numbercategory="22">Geography</button>
        </section>
    </aside>
  )
}