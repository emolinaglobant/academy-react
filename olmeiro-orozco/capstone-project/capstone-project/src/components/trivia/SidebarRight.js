import React from "react";
import { useSelector } from "react-redux";

export const SidebarRight = () => {
  const { totalPrize } = useSelector((state) => state.game);

  return (
    <aside className="trivia__sidebar-right">
      <h2>Outcomes:</h2>
      <section className="trivia__acumulated">
        <h4>Points accumulated:</h4>
        <p>{`${totalPrize}`}</p>
      </section>
      <section className="trivia__totalAward">
        <h4>Total Prize: </h4>
        <p>25 points</p>
      </section>
    </aside>
  );
};
