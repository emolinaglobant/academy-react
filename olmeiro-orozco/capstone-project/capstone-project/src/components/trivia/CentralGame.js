import React from "react";
import { CardQuestion } from "./CardQuestion";

export const CentralGame = () => {
  return (
    <section className="trivia__central animate__animated animate__backInUp">
      <CardQuestion />
    </section>
  );
};