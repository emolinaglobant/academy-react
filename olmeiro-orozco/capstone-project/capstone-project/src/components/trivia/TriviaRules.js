import React from "react";
import { Link } from "react-router-dom";
import { FooterGame } from "./FooterGame";

export const TriviaRules = () => {
  return (
    <article className="rules__container animate__animated animate__flash">
      <header>
        <h1>Reglas Trivia Game</h1>
      </header>

      <section className="rules__list">
        <ol>
          <li>Por cada pregunta bien contestada obtiene un punto.</li>
          <li>Puede retirarse antes sí lo desea con su acumulado.</li>
          <li>Si falla con la respuesta pierde el acumulado.</li>
          <li>Total número preguntas 25</li>
        </ol>

        <Link to="/" className="rules__link">
          Volver A TriviaGame
        </Link>
      </section>

      <footer>
        <FooterGame />
      </footer>
    </article>
  );
};