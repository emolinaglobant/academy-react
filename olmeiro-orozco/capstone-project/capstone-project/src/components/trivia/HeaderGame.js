import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { startLogout } from "../../actions/auth";
import { HeaderButtons } from "./HeaderButtons";

export const HeaderGame = () => {
  const { name } = useSelector((state) => state.auth);

  const imagePath = "/assets/trivia2.jpg";

  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <>
      <header className="trivia__header">
        <img
          src={imagePath}
          className="trivia__img animate__animated animate__flash"
          alt="trivia game"
        />
        <h1 className="animate__animated animate__bounce">
          Trivia Game: {name}
        </h1>
        <Link to="/rules" className="trivia__header-link">
          Rules
        </Link>
        <button
          className="btn btn__end animate__animated animate__rubberBand"
          onClick={handleLogout}
        >
          Logout
        </button>
      </header>
      <section className="trivia__header-buttons">
        <HeaderButtons />
      </section>
    </>
  );
};
