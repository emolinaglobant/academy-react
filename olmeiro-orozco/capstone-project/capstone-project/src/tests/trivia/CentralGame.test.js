import { shallow } from "enzyme"
import { CentralGame } from "../../components/trivia/CentralGame"

describe('Test component CentralGame', () => {
  test('should render component correctly', () => { 
    const wrapper = shallow(<CentralGame />);
    expect(wrapper).toMatchSnapshot();
   })
})