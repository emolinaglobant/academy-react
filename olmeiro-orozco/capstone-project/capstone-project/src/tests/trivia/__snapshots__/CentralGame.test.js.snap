// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component CentralGame should render component correctly 1`] = `
<section
  className="trivia__central animate__animated animate__backInUp"
>
  <CardQuestion />
</section>
`;
